---
title: "Validating Models in Clinical Data"
output: html_notebook
---


# Model design

For this section, we will be investigating validating a model in clincial data. 

```{r loadData}
library(PharmacoGx)
library(PharmacoGxML)
library(readr)
library(caret)
library(Biobase)


load("~/Code/pSetUploadR/PSets/CCLE.CTRPv2.RData")

library(GenomicFeatures)
txdb <- loadDb("~/Downloads/DEworkshop-master/transcriptome/gencode.v29.annotation.txdb")

tx2gene <-  GenomicFeatures::transcriptsBy(txdb)

tx2gene <- GenomicRanges::as.data.frame(tx2gene)[,c("tx_name","group_name")]

tx2gene[,1] <- gsub("\\.[0-9]*", rep="", x=tx2gene[,1])


library(readr)

mmrf_exp <- as.data.frame(read_tsv("mmrf_data/MMRF_CoMMpass_IA13a_E74GTF_Salmon_V7.2_Filtered_Transcript_TPM.txt"))
rownames(mmrf_exp) <- mmrf_exp[,1]
mmrf_exp <- mmrf_exp[,-1]

# 
# tximport::summarizeToGene(list(abundance=mmrf_exp[,1:5]), tx2gene=tx2gene,
#                           ignoreTxVersion=TRUE, countsFromAbundance="scaledTPM")
# 

library(PharmacoGxML)

CCLE.exp <- exprs(summarizeMolecularProfiles(CCLE.CTRPv2, "rnaseq", fill.missing = FALSE))
CTRPv2.aac <- summarizeSensitivityProfiles(CCLE.CTRPv2, drugs="Bortezomib", fill.missing = FALSE)

common.cells <- intersect(colnames(CCLE.exp),names(CTRPv2.aac))

CCLE.exp <- CCLE.exp[,common.cells]
CTRPv2.aac <- CTRPv2.aac[common.cells]


# 
# blood.cells <- intersect(common.cells, cellNames(CCLE.CTRPv2)[cellInfo(CCLE.CTRPv2)$tissueid == "haematopoietic_and_lymphoid_tissue"])
# 
# CCLE.exp <- CCLE.exp[,blood.cells]
# CTRPv2.aac <- CTRPv2.aac[blood.cells]

CCLE.exp <- CCLE.exp[complete.cases(CCLE.exp),] 

load("GeneSets-GO-KEGG-REACTOME.RData")


stress_pathways <- gTogs[grepl(gTogs[,2], pat="stress",ignore.case = T),]

stress.genes <- rownames(featureInfo(CCLE.CTRPv2, "rnaseq"))[which(featureInfo(CCLE.CTRPv2, "rnaseq")$EntrezGeneId %in% stress_pathways[,1])]

CCLE.exp <- CCLE.exp[stress.genes,]

test <- optimization(t(CCLE.exp),t(data.frame("Bortezomib" = CTRPv2.aac)), feature.selection = "variance", method="random_forest", features.no = nrow(CCLE.exp))


# 
# as.data.frame(read_tsv("landmark_genes.txt"))


```
